Southern University of Science and Technology-Department of Computer Science and Engineering

Course: Machine Learning(CS 405)-Professor: Qi Hao

## Homework #2
#### Due date: October, 7th, 2019

In [1]:
"""
Import libraries that you might require.

"""

import numpy as np
import math
import matplotlib.pyplot as plt
import operator

from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier

We will implement the KNN algorithm for the breast cancer dataset. Refer to the pdf and the following functions for the instructions. Complete all the functions as indicated below. The four functions would be autograded as mentioned in the pdf.

In [12]:
"""
Task 1: Classification

Please implement KNN for K: 3, 5, and 7 with the following norms:
L1
L2
L-inf
"""

# Read data (Breast Cancer Dataset). Remember to comment out the code not contained in a function.
from sklearn.datasets import load_breast_cancer

breast = load_breast_cancer()

X = breast['data']
y = breast['target']

np.random.seed(100)
p = np.random.permutation(len(X))
X, y = X[p], y[p]

X_train, y_train = X[:400], y[:400]
X_val, y_val = X[400:500], y[400:500]
X_test, y_test = X[500:], y[500:]


def distanceFunc(metric_type, vec1, vec2):
    """
    Computes the distance between two d-dimension vectors. 
    
    Please DO NOT use Numpy's norm function when implementing this function. 
    
    Args:
        metric_type (str): Metric: L1, L2, or L-inf
        vec1 ((d,) np.ndarray): d-dim vector
        vec2 ((d,)) np.ndarray): d-dim vector
    
    Returns:
        distance (float): distance between the two vectors
    """

    diff = vec1 - vec2
    if metric_type == "L1":
        distance = np.sum(vec1-vec2)
        #distance = 0 #complete

    if metric_type == "L2":
        distance = np.sqrt(np.sum(np.square(vec1-vec2)))
        #distance = 0 #complete
        
    if metric_type == "L-inf":
        distance = np.max(vec1-vec2)
        #distance = 0 #complete
        
    return distance


def computeDistancesNeighbors(K, metric_type, X_train, y_train, sample):
    """
    Compute the distances between every datapoint in the train_data and the 
    given sample. Then, find the k-nearest neighbors.
    
    Return a numpy array of the label of the k-nearest neighbors.
    
    Args:
        K (int): K-value
        metric_type (str): metric type
        X_train ((n,p) np.ndarray): Training data with n samples and p features
        y_train : Training labels
        sample ((p,) np.ndarray): Single sample whose distance is to computed with every entry in the dataset
        
    Returns:
        neighbors (list): K-nearest neighbors' labels
    """

    # You will also call the function "distanceFunc" here
    # Complete this function
    distances = []
    for i in range(X_train.shape[0]):
        dist = distanceFunc(metric_type, X_train[i], sample)
        distances.append((i, dist))
    distances.sort(key=operator.itemgetter(1))
    neighbors = []
    for j in range(K):
        neighbors.append(y_train[distances[j][0]])
    return neighbors


def Majority(neighbors):
    """
    Performs majority voting and returns the predicted value for the test sample.
    
    Since we're performing binary classification the possible values are [0,1].
    
    Args:
        neighbors (list): K-nearest neighbors' labels
        
    Returns:
        predicted_value (int): predicted label for the given sample
    """
    
    # Performs majority voting
    # Complete this function
    classVotes = {}
    for i in range(len(neighbors)):
        response = neighbors[i]
        if response in classVotes:
            classVotes[response] += 1
        else:
            classVotes[response] = 1
    sortedVotes = sorted(classVotes.items(), key=operator.itemgetter(1), reverse=True)
    predicted_value=sortedVotes[0][0]
    return predicted_value


def KNN(K, metric_type, X_train, y_train, X_val):
    """
    Returns the predicted values for the entire validation or test set.
    
    Please DO NOT use Scikit's KNN model when implementing this function. 

    Args:
        K (int): K-value
        metric_type (str): metric type
        X_train ((n,p) np.ndarray): Training data with n samples and p features
        y_train : Training labels
        X_val ((n, p) np.ndarray): Validation or test data
        
    Returns:
        predicted_values (list): output for every entry in validation/test dataset 
    """
    
    # Complete this function
    # Loop through the val_data or the test_data (as required)
    # and compute the output for every entry in that dataset  
    # You will also call the function "Majority" here
    predictions = []
    for i in range(X_val.shape[0]):
        neighbors = computeDistancesNeighbors(K, metric_type, X_train, y_train, X_val[i])
        predicted_value = Majority(neighbors)
        predictions.append(predicted_value)
    return predictions


def evaluation(predicted_values, actual_values):
    """
    Computes the accuracy of the given datapoints.
    
    Args:
        predicted_values ((n,) np.ndarray): Predicted values for n samples
        actual_values ((n,) np.ndarray): Actual values for n samples
    
    Returns:
        accuracy (float): accuracy
    """
    
    return accuracy_score(predicted_values, actual_values)


def main():
    """
    Calls the above functions in order to implement the KNN algorithm.
    
    Test over the following range K = 3,5,7 and all three metrics. 
    In total you will have nine combinations to try.
    
    PRINTS out the accuracies for the nine combinations on the validation set,
    and the accuracy on the test set for the selected K value and appropriate norm.
    
    REMEMBER: You have to report these values by populating the Table 1.
    """
    
    ## Complete this function
    
    K = [3,5,7]
    norm = ["L1", "L2", "L-inf"]
    for K in [3,5,7]:
        for norm in ["L1", "L2", "L-inf"]:
            print("<<<<VALIDATION DATA PREDICTIONS>>>>")
            predicted_values = KNN(K, norm, X_train, y_train, X_val)
            print("K@{}, norm@{}, Accuracy:{}".format(K, norm, evaluation(predicted_values, y_val)))
            ## Complete
            print("<<<<TEST DATA PREDICTIONS>>>>")
            predicted_values = KNN(K, norm, X_train, y_train, X_test)
            print("K@{}, norm@{}, Accuracy:{}".format(K, norm, evaluation(predicted_values, y_test)))
            ## Complete

Uncomment the code below to run the main function (Remember to recomment the code before submitting).

In [13]:
# Finally, call the main function
main()

<<<<VALIDATION DATA PREDICTIONS>>>>
K@3, norm@L1, Accuracy:0.6
<<<<TEST DATA PREDICTIONS>>>>
K@3, norm@L1, Accuracy:0.5652173913043478
<<<<VALIDATION DATA PREDICTIONS>>>>
K@3, norm@L2, Accuracy:0.95
<<<<TEST DATA PREDICTIONS>>>>
K@3, norm@L2, Accuracy:0.8840579710144928
<<<<VALIDATION DATA PREDICTIONS>>>>
K@3, norm@L-inf, Accuracy:0.6
<<<<TEST DATA PREDICTIONS>>>>
K@3, norm@L-inf, Accuracy:0.5942028985507246
<<<<VALIDATION DATA PREDICTIONS>>>>
K@5, norm@L1, Accuracy:0.6
<<<<TEST DATA PREDICTIONS>>>>
K@5, norm@L1, Accuracy:0.5652173913043478
<<<<VALIDATION DATA PREDICTIONS>>>>
K@5, norm@L2, Accuracy:0.93
<<<<TEST DATA PREDICTIONS>>>>
K@5, norm@L2, Accuracy:0.8985507246376812
<<<<VALIDATION DATA PREDICTIONS>>>>
K@5, norm@L-inf, Accuracy:0.6
<<<<TEST DATA PREDICTIONS>>>>
K@5, norm@L-inf, Accuracy:0.5652173913043478
<<<<VALIDATION DATA PREDICTIONS>>>>
K@7, norm@L1, Accuracy:0.6
<<<<TEST DATA PREDICTIONS>>>>
K@7, norm@L1, Accuracy:0.5652173913043478
<<<<VALIDATION DATA PREDICTIONS>>>>
K@7, 

Answer the following questions here:

1. How could having a larger dataset influence the performance of KNN?

2. Tabulate your results from `main()` in the table provided.

3. Finally, mention the best K and the norm combination you have settled upon and report the accuracy on the test set using that combination.

2.1Validate Set
K   Norm   Accuracy
3   L1    0.6
3   L2    0.95
3   L-Inf  0.6 
5   L1    0.6
5   L2    0.93
5   L-Inf  0.6 
7   L1    0.6
7   L2    0.92
7   L-Inf  0.6 
2.2 Test Set
K   Norm   Accuracy
3   L1    0.56
3   L2    0.88
3   L-Inf  0.59 
5   L1    0.56
5   L2    0.89
5   L-Inf  0.56 
7   L1    0.56
7   L2    0.91
7   L-Inf  0.56 

3.For validate set, the best combination is k=3 and norm =L2 , and the test set achieve accuracy=0.88 which is not the best performance. There exists little over-fitting.